### ISSUE TITLE
Visualizing attention weights?

### ISSUE THREAD

**Issue by justindujardin** [lucidrains/reformer-pytorch#27]
Thanks for the cool library! 

I'm working on a seq2seq demo using it, and I'd like to visualize the attention weights, but it isn't clear how to get them out of the ReformerLM class. Can you point me in the right direction?

**Comment by lucidrains** [comment]
Thanks Justin! Are you interested in the final unsorted attention logits? I can think of a good way to expose that (along with the bucket distribution), if that's what you'd like

**Comment by justindujardin** [comment]
I guess so? I was hoping to integrate the weights with this visualizer: https://github.com/jessevig/bertviz

It says that it wants:
```
attention: list of `torch.FloatTensor`(one for each layer) of shape
    `(batch_size(must be 1), num_heads, sequence_length, sequence_length)`
```

**Comment by lucidrains** [comment]
gotcha, I can expose that!

**Comment by lucidrains** [comment]
@justindujardin I've made an attempt to unsort the attention weights of the buckets. you can find the code here https://github.com/lucidrains/reformer-pytorch/commit/873a26a64b4dfac3677aa32083ed3979475d2d02 It was quite tricky and I'm still not 100% sure if it is done correctly or that it is bug free. I am also using a third party library https://github.com/rusty1s/pytorch_scatter, so I would recommend following their instructions for setting up CUDA if you run into any errors when turning on the functionality I just built.

To try it out, please follow the instructions here https://github.com/lucidrains/reformer-pytorch#research I'll keep this issue open until you've gotten a chance to test it.

**Comment by pabloppp** [comment]
Hi @lucidrains thanks for your awesome work!
When I wrap the model in the Recorder class I get the following error:
<img width="495" alt="Captura de pantalla 2020-02-10 a las 8 59 51" src="https://user-images.githubusercontent.com/4237014/74131135-f814e600-4be3-11ea-92bb-6be9e4cbd2a1.png">

Looks like maybe "modules" is not a function, but it's the module list itself?

**Comment by lucidrains** [comment]
@pabloppp hey no problem! i found a bug lol, could you try upgrading and trying again?

**Comment by pabloppp** [comment]
I now seem to be able to display the attention, thank you!
But for some reason, it looks like the model is masking the current element (the query) when performing attention.

<img width="246" alt="Captura de pantalla 2020-02-10 a las 10 30 26" src="https://user-images.githubusercontent.com/4237014/74137734-9dce5200-4bf0-11ea-9169-1a303fae0141.png">

I am attempting to do 2D attention, and this is why it's displayed as a 2D image but it looks like the pixel used as a query in the attention mechanism has 0 attention 🤔I think this should not be the case, as a word (or a pixel in my case) should be able to have attention on itself, right?

The attention that I show there is from an untrained model, I will attempt to train it and see if I am able to see something cooler :D

**Comment by lucidrains** [comment]
Nice! So the token not being to attend to itself is actually something this paper proposes is necessary for LSH to work. Since queries and keys must share the same space, a token paying attention to itself would not make sense. So they mask it out to some arbitrarily low number. You should continue your experiments and see what you find though!

**Comment by lucidrains** [comment]
Yeah, fetching the full attention matrix is expectedly costly. I'll see what I can do to lower the memory usage this week

**Comment by pabloppp** [comment]
Thank you very much for your effort @lucidrains !!!

**Comment by JorgeCeja** [comment]
Hey @justindujardin, I'm also working on seq2seq experiments. This is off topic but I was wondering if you would be able to share your demo? I'm having trouble decoding the outputs (sampling) for inference. Thanks!

**Comment by lucidrains** [comment]
Ok, I've made the calculation and recording of the full attention matrix a bit more memory efficient with https://github.com/lucidrains/reformer-pytorch/commit/631a4efde06fe06e26aae1e394e522d93c121315

**Comment by justindujardin** [comment]
I was able to use the exported attention data and see the expected patterns. I'll look at integrating it with the linked viz tool over the weekend. Thanks for the excellent turnaround time and awesome work! 🏆 👏 

@JorgeCeja I'm not using a decoder, but I'll put up this demo somewhere once I clean things up. Perhaps @lucidrains would expand on the seq2seq encode/decoder sample to add a snippet that uses the decoder for sampling if you opened a new issue?
